### PR TITLE
Handle gomod file paths safely

### PR DESCRIPTION
**Pull request by chmeliik**
# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [x] New code has type annotations
- [n/a] Docs updated (if applicable)
- [n/a] Docs links in the code are still valid (if docs were updated)

### CODE REVIEW

File: cachi2/core/errors.py [hunk 1]
```diff
@@ -1,107 +1,100 @@
...
+            self.solution = self.default_solution
+        else:
+            self.solution = solution
+        self.docs = docs
 
     def friendly_msg(self) -> str:
         """Return the user-friendly representation of this error."""
-        return str(self)
+        msg = str(self)
+        if solution := self.solution:
```
Comment (lkolacek) [inline cachi2/core/errors.py:36]: `if self.solution:` would be probably sufficient here. :)

Comment (chmeliik) [reply cachi2/core/errors.py]: True, I was just too lazy to type `self.` twice when moving the code out of the function :smile:

**Review by lkolacek — APPROVED** [review]
Looks great. Thank you, Adam! Coverage dropped by 0.05% which is in my opinion tolerable.

**Comment by chmeliik** [comment]
> Looks great. Thank you, Adam! Coverage dropped by 0.05% which is in my opinion tolerable.

I checked and realized that there was no unit test for the pydantic integration of RootedPath :facepalm: 

added

**Comment by brunoapimentel** [comment]
Great final solution with `RootedPath` :rocket::rocket::rocket:

So, what how are we planning to tackle the changes to the pip part? Are we doing it in a follow up PR, or are we filing a ticket to handle it later?

**Comment by chmeliik** [comment]
> Great final solution with `RootedPath` rocketrocketrocket
> 
> So, what how are we planning to tackle the changes to the pip part? Are we doing it in a follow up PR, or are we filing a ticket to handle it later?

Give me a minute

**Comment by chmeliik** [comment]
pip: #126

File: cachi2/core/package_managers/gomod.py [hunk 2]
```diff
@@ -136,7 +137,33 @@ def fetch_gomod_source(request: Request) -> RequestOutput:
...
+    where it is executed. Check some of the common paths that the subcommands may read.
+    """
+
+    def check_potential_symlink(relative_path: Union[str, Path]) -> None:
+        try:
+            app_dir.join_within_root(relative_path)
+        except PathOutsideRoot:
+            raise PackageRejected(
+                reason=(
+                    f"Found suspicious symlink ({relative_path}) in Go module directory "
```
Comment (brunoapimentel) [inline cachi2/core/package_managers/gomod.py:153]: I think we should explain briefly in this exception why the symlink is suspicious (i.e. points to ouside of the repo).

**Review by brunoapimentel — APPROVED** [review]

**Comment by brunoapimentel** [comment]
I noticed we have an unused function in the gomod code: https://github.com/containerbuildsystem/cachi2/blob/main/cachi2/core/package_managers/gomod.py#L713-L716

It does not have anything to do with your PR, but if since you changed that file, if you could kindly remove it so we don't need to do it in a standalone PR :pleading_face:

File: cachi2/core/package_managers/gomod.py (hunk 2 above)
Comment (chmeliik) [reply cachi2/core/package_managers/gomod.py]: I decided to raise the original PathOutsideRoot and just modify the "solution" to explain the symlink problem

Comment (chmeliik) [reply cachi2/core/package_managers/gomod.py]: @brunoapimentel PTAL if it's better now

Comment (brunoapimentel) [reply cachi2/core/package_managers/gomod.py]: Looks great :+1:

**Pull request merged**